### ISSUE TITLE
Does an LDA-based binary classifier make sense for dimensionality reduction?

### ISSUE THREAD

**Issue by tomaz-suller** [zuhairmhtb/P300SpellerBasedEEGClassification#2]
This issue is meant to be more of a discussion, since I have to say I'm not familiar with the statistics behind linear discriminant analysis. Still, the [scikit-learn documentation](https://scikit-learn.org/stable/modules/lda_qda.html#dimensionality-reduction-using-linear-discriminant-analysis) seems to advise against using LDA for dimensionality reduction under a binary classification problem, like the one we face here (emphasis is mine):

> The dimension of the output is necessarily less than the number of classes, so this is in general a rather strong dimensionality reduction, and **only makes sense in a multiclass setting**.

Considering we do use this technique (like the notebook in this repository does), we're limiting our model to classifying based on a single feature, which clearly should make the model fragile to any slight variation in its input, which is especially frequent in the BCI application domain. Still, both during cross validation and testing, providing only this single variable to Random Forest allows for apparently good performance, with >80% accuracy.

What gives? I believe I could either be missing some theory about the domain (which is likely) or that there might be some issue with some step during modelling in which there was a data leak or something like that.

**Comment by zuhairmhtb** [comment]
Hello, Sorry for the late reply and thank you for pointing it out. I agree that using a single feature for classification is not ideal. Regarding LDA - the paper (https://www.frontiersin.org/articles/10.3389/fnhum.2013.00732/full) used stepwise LDA for classification and as a result I tested my data with LDA. I also had a feeling that the high accuracy may have occurred due to data leak. Therefore, I have done the same processing but slightly with a different approach. You can find the changes in EEGAnalysis.ipynb file. However, I still get an accuracy over 80% when I use RFC with LDA. 

You will notice that in the visualization, the separation of data is not very clear when I use the two LDA features. It is more or less homogeneous since the first feature (along x-axis) that has the highest variance only contributes a little to the separation. But even with this homogeneous distribution, RFC does a good job of classifying target and non-target signals.

Please let me know if you spot any error that might be leading to this result. It has been years since I have written this code and it is very possible that I may have overlooked something important. I will perform a deeper analysis on this and I will post it here if I find any solution to why the accuracy with this approach of classification is high.

PS: If the code is correct, one hypothesis is - the train and test data is different but both contains samples from all patients. Since all samples from each patients were collected during the same experiment, the samples may have some kind of similarity between them and probably RFC is picking it up and is using it to predict the test data. I will test it later by splitting the train and test data in such a way that test data contains samples from patients on which RFC was not trained.

**Edited: Found the reason. The dataset was not balanced. After balancing the dataset, the accuracy reduced to 60% which looks more accurate.**
